### PR TITLE
Alter mda method

### PR DESCRIPTION
**Pull request by mattg3004**
Align the MDA method with other diseases using beta random variables for probability of treatment based on coverage and snc values. When coverage or snc change, the probability of treatment will change for everyone, but the rank of probability will stay the same, e.g. the person who was most likely to get treated with the previous coverage and snc will still be the person most likely to get treated after the change.

When the population is initialized everyone is assigned a "treatProbability" of -1. At the first MDA, this value is updated for everyone. When someone dies, this is set to -1 again.  At subsequent MDA's, if there is a different value of coverage or snc from the previous MDA then we first check for anyone with a -1 value, so that we can give them a value for treatProbability with the old coverage and snc, then they can be properly ranked alongside everyone else.

Requires the parameter file to have an "snc" parameter now, as this is read in when loading parameters.

### CODE REVIEW

File: sch_simulation/helsim_FUNC_KK/utils.py [hunk 1]
```diff
@@ -419,3 +419,58 @@ def getPsi(params: Parameters) -> float:
...
+            # Draw probabilities from the beta distribution for each persons probability of being traeted
+            SD.treatProbability[notInitTreatmentProb] = np.random.beta(alpha, beta, len(notInitTreatmentProb))
+        else:
+            SD.treatProbability[notInitTreatmentProb] = np.ones(len(notInitTreatmentProb)) * cov           
+    return SD
+
+
+
+def editTreatProbability(SD: SDEquilibrium, cov: float, snc: float) -> SDEquilibrium:
+    # we want to choose new values for treatment probability when the coverage or snc change
```
[the remark this answers is not in the document]
Comment (thk123) [reply sch_simulation/helsim_FUNC_KK/utils.py:452]: These could be added to the other two, I think `checkForZeroTreatProbability` is for picking up those who've died and giving their replacements a probability, and `initializeTreatmentProbability` is for running at the start (though wouldn't check do that too?)

**Review by thk123 — COMMENTED** [review]
I have some design thoughts but I'm also tempted to hold out till NTD-Modelling-Consortium/endgame-project#35 so we can write some unit tests - I'd be delighted to help, this feels like perfect feature to demonstrate the key ideas. The only question is whether we need to kick off the fitting with this sooner than we can get that all sorted.

File: sch_simulation/helsim_FUNC_KK/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mattg3004) [reply sch_simulation/helsim_FUNC_KK/utils.py:452]: 💚 you're right, the initialization function isn't needed

File: sch_simulation/helsim_RUN_KK.py [hunk 2]
```diff
@@ -539,13 +545,29 @@ def doRealizationSurveyCoveragePickle(
                 
             # chemotherapy
             if timeBarrier >= nextChemoTime:
-                #print("MDA, time = ", t)           
+                
+                if nChemo == 0: # if this is the first MDA, then we need to initialize everyone's probability of treatment
```
[the remark this answers is not in the document]
Comment (mattg3004) [reply sch_simulation/helsim_RUN_KK.py:549]: If we do this, then we have to use the MDA data and store the previously used coverage and snc values. Seemed better to just do it later if needed to me, but can be changed if needed

File: sch_simulation/helsim_FUNC_KK/configuration.py [hunk 3]
```diff
@@ -201,7 +201,8 @@ def setupSD(params: Parameters) -> SDEquilibrium:
         numSurvey=0,
         nChemo1 = 0,
         nChemo2 = 0, 
-        id = ids
+        id = ids,
+        treatProbability = np.full(shape=params.N, fill_value=np.NaN, dtype=float)
```
Comment (tlestang) [inline sch_simulation/helsim_FUNC_KK/configuration.py:205]: Default value is set to `np.NaN` here, but to `-1` in `helsim_RUN_KK.multiple_simulations` (e.g. [here](https://github.com/NTD-Modelling-Consortium/ntd-model-sch/pull/45/files#diff-81f0d3c4740bc6658d8756d85ec3a3a89dfa112cdab42c389febaaaf4d541ef0R1038)).

In addition it looks to me like this function `setupSD` initialising the treatment probability to Nan is only called from `helsim_RUN_KK.doRealizationSurveyCoveragePickle`

```python
def doRealizationSurveyCoveragePickle(..., simData):
    # ...
    # simData is not None and simData.treatProbability and default probabiltity of tment is -1
    if simData is None:
        simData = setupSD(params)  # Sets simData.treatProbability to np.NaN
```

**Comment by tlestang** [comment]
I see you went for an implementation very similar to the one featured in LF, where you use a placeholder value for the treatment probability of reset individuals until this value is computed later at the time of the next MDA. I think I get the logic, but I don't understand why not computing the next treatment probability value upon reset in `doDeath` directly.

Something like:

```python
# helsim_FUNC_KK/events.py
def doDeath(..., params, ...):
    # ...
    new_ids = np.arange(maxID + 1, maxID + len(theDead) + 1)
    SD.id[theDead] = new_ids
    cov = get_coverage(params)
    SD.treatProbability[theDead] = drawTreatmentProbabilities(
        len(new_ids), cov=cov, snc=params.snc,
    )
```

then when the next MDA comes with different coverage and snc values, you only have to compute new probability values and order them accordingly to ensure rank correlation. Am I missing something? 

It's the way it was implemented in Oncho (https://github.com/NTD-Modelling-Consortium/EPIONCHO-IBM/pull/60)

File: sch_simulation/helsim_FUNC_KK/configuration.py (hunk 3 above)
Comment (mattg3004) [reply sch_simulation/helsim_FUNC_KK/configuration.py:205]: I've amended the first point, thanks for the catch

the setupSD is called when we want to generate a new simulation when we don't have a stored population to start from. It does more than is included in the comment, as it generates the rest of the data about the population at the starting point of the simulation.

**Comment by thk123** [comment]
Just noting the CI job that is failing is failing because this repo wrongly installs numpy2.0 - I'll get my fix for this into its own PR that we can merge ASAP.

**Comment by thk123** [comment]
I've now fixed the numpy issue on master, so if you merge that into this branch the CI should go green.

File: sch_simulation/helsim_FUNC_KK/utils.py [hunk 4]
```diff
@@ -420,45 +420,38 @@ def getPsi(params: Parameters) -> float:
...
 
-def initializeTreatmentProbability(SD: SDEquilibrium, cov: float, snc: float) -> SDEquilibrium:
 
-    if(snc > 0):
+def drawTreatmentProbabilities(n:int, cov:float, snc:float):
+
+    """
+    Draw the treatment probabilities for the value of coverage and snc given.
+    This uses the scheme explained in section 1.5.3 of the suppplement to this paper
+    https://www.sciencedirect.com/science/article/pii/S1755436516300810?via%3Dihub#sec0110
```
Comment (thk123) [inline sch_simulation/helsim_FUNC_KK/utils.py:429]: 💚 Great docs!

File: sch_simulation/helsim_FUNC_KK/utils.py [hunk 5]
```diff
@@ -420,45 +420,38 @@ def getPsi(params: Parameters) -> float:
...
+    """
+
+    if(cov == 0):
```
Comment (thk123) [inline sch_simulation/helsim_FUNC_KK/utils.py:432]: 💚 This whole function seems really clear to me - thanks for pulling it out.

File: sch_simulation/helsim_RUN_KK.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (thk123) [reply sch_simulation/helsim_RUN_KK.py]: I don't fully understand this I'm afraid - but the answer seems clearly there is a reason so let's leave it as is for now and maybe you can explain this more fully next time we're chatting on a call

**Review by thk123 — APPROVED** [review]
Thanks for addressing all of my key issues.

**Pull request merged**